File: 1392.最长快乐前缀.js
/*
 * @lc app=leetcode.cn id=1392 lang=javascript
 *
 * [1392] 最长快乐前缀
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPrefix = function (s) {
    for (let offset = 1; offset < s.length; offset++) {
        for (let start = 0; s[start] === s[start + offset]; start++) {
            if (start + offset === s.length - 1) return s.substring(offset);
        }
    }
    return '';
};
// @lc code=end

File: 1392.最长快乐前缀.java
/*
 * @lc app=leetcode.cn id=1392 lang=java
 *
 * [1392] 最长快乐前缀
 */

// @lc code=start
class Solution {
    public String longestPrefix(String s) {
        int n = s.length();
        long prefix = 0;
        long suffix = 0;
        long base = 31;
        long mul = 1;
        long mod = 1000000007;
        int happy = 0;

        for (int i = 1; i < n; i++) {
            prefix = (prefix * base + (s.charAt(i - 1) - 'a')) % mod;
            suffix = (suffix + (s.charAt(n - i) - 'a') * mul) % mod;
            if (prefix == suffix) {
                happy = i;
            }
            mul = mul * base % mod;
        }

        return s.substring(0, happy);
    }
}
// @lc code=end
